servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java

```diff
@@ -61,15 +112,66 @@ public static synchronized void initializeFileSystem(boolean permissionCheckEnab
    * @param inodes all the inodes retrieved by traversing the path
    * @throws IOException if permission checking fails
    */
-  public static void checkPermission(User user, List<String> groups, FileSystemAction action,
+  private static void checkOnInodes(String user, List<String> groups, FileSystemAction action,
       List<Inode> inodes) throws IOException {
+    int size = inodes.size();
+    assert size > 0;
+
     if (!sPermissionCheckEnabled) {
       return;
     }
 
-    if (sFileSystemOwner.equals(user.getName()) || groups.contains(sFileSystemSuperGroup)) {
+    // bypass checking permission for super user or super group of Tachyon file system.
+    if (sFileSystemOwner.equals(user) || groups.contains(sFileSystemSuperGroup)) {
+      return;
+    }
+
+    // traverses parent path to ensure inodes in it are all executable
+    for (int i = 0; i < size - 1; i++) {
+      check(user, groups, inodes.get(i), FileSystemAction.EXECUTE);
+    }
+
+    check(user, groups, inodes.get(size - 1), action);
+  }
+
+  /**
+   * This method check requested permission on a given inode.
+   *
+   * @param user who requests access permission
+   * @param groups in which user belongs to
+   * @param inode on which apply permission check logic
+   * @param action requested {@link FileSystemAction} by user
+   * @throws IOException if permission checking fails
+   */
+  private static void check(String user, List<String> groups, Inode inode, FileSystemAction action)
```

Comment:
Also, it's discouraged to expose `Inode` directly to outside, since it's the core metadata to the file system. Is it possible to get `FileInfo` from master, and use it for permission check?